File: training/Leetcode/23.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None
from Queue import PriorityQueue
class Solution(object):
    def mergeKLists(self, lists):
        """
        :type lists: List[ListNode]
        :rtype: ListNode
        """
        head = cur = ListNode(0)
        que = PriorityQueue()
        for node in lists:
            if node:
                que.put((node.val, node))
        while not que.empty():
            val, node = que.get()
            cur.next = ListNode(val)
            cur = cur.next
            node = node.next
            if node:
                que.put((node.val, node))
        return head.next

File: training/Leetcode/23.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        auto compare = [](ListNode* l, ListNode* r) {
            return l->val > r->val;
        };
        priority_queue<ListNode*, vector<ListNode*>, decltype(compare)> que(compare);

        for (int i = 0; i < lists.size(); i++) {
            if (lists[i] != NULL)
                que.push(lists[i]);
        }

        ListNode *list = NULL, *cur = NULL;
        while (que.size()) {
            ListNode* curNode = que.top(); que.pop();

            if (list == NULL) {
                list = curNode;
                cur = list;
            } else {
                cur->next = curNode;
                cur = cur->next;
            }
            if (curNode->next != NULL)
                que.push(curNode->next);
        }
        return list;
    }
};
